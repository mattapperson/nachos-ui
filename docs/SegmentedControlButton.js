import React, { Component } from "react";
import { View } from "react-native";

import Switcher from "../src/Switcher";
import SegmentedControlButton from "../src/SegmentedControlButton";

import B from "../src/typography/B";
import P from "../src/typography/P";
import H4 from "../src/typography/H4";

// WEBPACK (SegmentedControlButtonExample)
class SegmentedControlButtonExample extends Component {
  state = { valueOne: null, valueTwo: null };
  render() {
    return (
      <View>
        <H4 style={{ marginBottom: 20 }}>Example:</H4>

        <Switcher
          onChange={valueOne => this.setState({ valueOne })}
          direction="row"
        >
          <SegmentedControlButton
            value="volume"
            text="Volume"
            iconName="md-volume-off"
          />
          <SegmentedControlButton value="walk" text="Walk" iconName="md-walk" />
          <SegmentedControlButton value="wine" text="Wine" iconName="md-wine" />
        </Switcher>

        <P style={{ marginVertical: 20 }}>
          <B>Selected value:</B> {this.state.valueOne || "no selection"}
        </P>

        <Switcher
          onChange={valueTwo => this.setState({ valueTwo })}
          direction="column"
        >
          <SegmentedControlButton
            value="volume"
            text="Volume"
            iconName="md-volume-off"
          />
          <SegmentedControlButton value="walk" text="Walk" iconName="md-walk" />
          <SegmentedControlButton value="wine" text="Wine" iconName="md-wine" />
        </Switcher>

        <P style={{ marginVertical: 20 }}>
          <B>Selected value:</B> {this.state.valueTwo || "no selection"}
        </P>
      </View>
    );
  }
}
// WEBPACK

const description = `
**SegmentedControlButton** is just composition of the Button component. **You can pass all Button's props.**

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| \`value\` | \`string\` | \`void\` | Defines a value associated with the component  |
| \`text\` | \`string\` | \`void\` | Defines a text value for the Button |
| \`uppercase\` | \`bool\` | \`true\` | Determines whether text values should be uppercased  |
| \`theme\` | \`object\` | \`defaultTheme\` | An object describing a style of the component. See more in Theme section |
`;

SegmentedControlButtonExample.styleguide = {
  ...SegmentedControlButtonExample.styleguide,
  title: "SegmentedControlButton",
  description
};

export default SegmentedControlButtonExample;
