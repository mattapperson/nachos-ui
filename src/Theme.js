import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
