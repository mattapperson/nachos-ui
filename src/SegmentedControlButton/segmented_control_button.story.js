import { storiesOf } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import React from "react";
import SegmentedControlButton from "./index";
import Switcher from "../Switcher";
import { View } from "react-native";

const Example = () => {
  return (
    <View style={{ flex: 1, flexDirection: "column" }}>
      <Switcher direction="row">
        <SegmentedControlButton
          value="volume"
          text="Volume"
          iconName="md-volume-off"
        />
        <SegmentedControlButton value="walk" text="Walk" iconName="md-walk" />
        <SegmentedControlButton value="wine" text="Wine" iconName="md-wine" />
      </Switcher>
      <br />
      <br />
      <Switcher direction="column">
        <SegmentedControlButton
          value="volume"
          text="Volume"
          iconName="md-volume-off"
        />
        <SegmentedControlButton value="walk" text="Walk" iconName="md-walk" />
        <SegmentedControlButton value="wine" text="Wine" iconName="md-wine" />
      </Switcher>
    </View>
  );
};

storiesOf("SegmentedControlButton", module).add("default", withInfo()(Example));

export default Example;
