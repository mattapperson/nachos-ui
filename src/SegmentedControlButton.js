import React from "react";
import PropTypes from "prop-types";
import Button from "./Button";
import { withTheme } from "./Theme";

const SegmentedControlButton = props => {
  const {
    value,
    text,
    style,
    textStyle,
    // NOTE: injected by a Switcher
    direction,
    first,
    last,
    selected,
    onChange,
    iconSize,
    iconPosition,
    theme
  } = props;

  // NOTE: function onChange is injected by the Switcher component
  const switcherProp = onChange && {
    onPress: () => {},
    onPressOut: onChange.bind(null, value)
  };

  // NOTE: Clone props and then delete Component specific props so we can
  // spread the rest
  const { ...rest } = props;
  delete rest.direction;
  delete rest.first;
  delete rest.last;
  delete rest.value;
  delete rest.text;
  delete rest.onChange;
  delete rest.style;
  delete rest.textStyle;
  delete rest.theme;
  delete rest.kind;

  return (
    <Button
      iconColor={theme.baseText.color}
      iconActiveColor={theme.states.selectedText.color}
      iconSize={iconSize}
      iconPosition={iconPosition}
      kind="squared"
      {...rest}
      {...switcherProp}
      style={[
        theme.base,
        first ? theme.first[direction] : {},
        last ? theme.last[direction] : {},
        selected ? theme.states.selected : {},
        style
      ]}
      textStyle={[
        theme.baseText,
        selected ? theme.states.selectedText : {},
        textStyle
      ]}
    >
      {text}
    </Button>
  );
};

SegmentedControlButton.themeConfig = {
  settings: {
    iconSize: 20,
    iconPosition: "left"
  },
  style: {
    base: {
      height: 50,
      padding: 0,
      paddingHorizontal: 0,
      borderWidth: 1,
      borderColor: "#ddd",
      backgroundColor: "#fff"
    },
    baseText: {
      color: "#bdc1cc",
      fontSize: 12,
      fontWeight: "500"
    },
    states: {
      selected: {
        backgroundColor: "#2f8cff",
        borderColor: "#2f8cff"
      },
      selectedText: {
        color: "#fff"
      }
    },
    first: {
      row: {
        borderBottomLeftRadius: 0,
        borderTopLeftRadius: 0,
        borderRightWidth: 0
      },
      column: {
        borderTopLeftRadius: 0,
        borderTopRightRadius: 0,
        borderBottomWidth: 0
      }
    },
    last: {
      row: {
        borderBottomRightRadius: 0,
        borderTopRightRadius: 0,
        borderLeftWidth: 0
      },
      column: {
        borderBottomLeftRadius: 0,
        borderBottomRightRadius: 0,
        borderTopWidth: 0
      }
    }
  }
};

SegmentedControlButton.propTypes = {
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  text: PropTypes.string,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  textStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  uppercase: PropTypes.bool,
  theme: PropTypes.object,
  // NOTE: injected by a Switcher
  direction: PropTypes.oneOf(["row", "column"]),
  onChange: PropTypes.func,
  first: PropTypes.bool,
  last: PropTypes.bool,
  selected: PropTypes.bool
};

SegmentedControlButton.defaultProps = {
  uppercase: false
};

export default withTheme("SegmentedControlButton", SegmentedControlButton);
