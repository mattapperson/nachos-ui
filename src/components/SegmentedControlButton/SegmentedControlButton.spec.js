import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import SegmentedControlButton from "../SegmentedControlButton";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <SegmentedControlButton text="Feed" />
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
