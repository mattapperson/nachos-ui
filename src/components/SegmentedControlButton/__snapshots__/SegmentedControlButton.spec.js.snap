// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexGrow-16y2uox rn-flexShrink-1wbh5a2 rn-flexBasis-1ro0kt6 rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
>
  <div
    className="rn-alignItems-1oszu61 rn-backgroundColor-14lw9ot rn-borderTopColor-1npdrp0 rn-borderRightColor-aqtk88 rn-borderBottomColor-j9d2do rn-borderLeftColor-1jj1lyi rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-5kkj8d rn-borderRightWidth-13l2t4g rn-borderBottomWidth-qklmqi rn-borderLeftWidth-1ljd8xs rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-height-1iud8zs rn-justifyContent-1777fci rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-touchAction-1otgn73 rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr rn-zIndex-1lgpqti"
    data-focusable={true}
    disabled={false}
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="button"
    tabIndex="0"
  >
    <div
      className="rn-alignItems-1awozwy rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-18u37iz rn-justifyContent-1777fci rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
    >
      <div
        className="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-1471scf rn-fontFamily-baq3cm rn-fontStyle-13uqrnb rn-fontVariant-ebii48 rn-height-1pexk7n rn-lineHeight-rjixqe rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingBottom-1mi0q7o rn-paddingTop-wk8lta rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
        dir="auto"
        style={
          Object {
            "WebkitAlignSelf": "center",
            "alignSelf": "center",
            "color": "rgba(117,64,238,1.00)",
            "fontSize": "12px",
            "fontWeight": "500",
            "msFlexItemAlign": "center",
            "paddingLeft": "8px",
            "paddingRight": "8px",
            "textAlign": "center",
          }
        }
      >
        Feed
      </div>
    </div>
  </div>
</div>
`;
